File: src/observer_ffi.mjs
// @ts-check
/**
 * @template T
 * @typedef {(value: T) => void} Callback
 */

/**
 * @template T
 * @typedef  {{[index: number]: Callback<T>}} Callbacks
 */

/**
 * @template T
 * @typedef  {{[index: number]: [string[][], Callback<T>]}} TopicCallbacks
 */

/**
 * @template T
 * @typedef {{ indices: number[], next: TopicIndex<T>}} TopicIndexEntry
 */

/**
 * @template T
 * @typedef {{[topic: string]: TopicIndexEntry<T>}} TopicIndex
 */

/**
 * @template T
 * @typedef {{callbacks: Callbacks<T>, index: number}} Hub
 */

/**
 * @template T
 * @typedef {{callbacks: Callbacks<T>, index: number, value: T}} StatefulHub
 */

/**
 * @template T
 * @typedef {{callbacks: TopicCallbacks<T>, index: number, topicIndex: TopicIndex<T>}} TopicBasedHub
 */

// Helper functions
// ================

/**
 * Invoke each callback synchronously with the provided value.
 * @template T
 * @param {Callbacks<T>} callbacks
 * @param {T} value
 */
function invokeCallbacks(callbacks, value) {
  for (const [_, callback] of Object.entries(callbacks)) {
    callback.call(null, value);
  }
}

/**
 * Updates the topic index with a new subscription.
 * @template T
 * @param {TopicIndex<T>} topicIndex
 * @param {string[][]} topics
 * @param {number} index
 */
function updateTopicIndex(topicIndex, topics, index) {
  if (topics.length !== 0) {
    const [first, ...rest] = topics;

    for (const topic of first) {
      if (!topicIndex[topic]) {
        topicIndex[topic] = { indices: [], next: {} };
      }
      if (rest.length === 0) {
        topicIndex[topic].indices.push(index);
      } else {
        updateTopicIndex(topicIndex[topic].next, rest, index);
      }
    }
  }
}

/**
 * Removes a subscription from the topic index.
 * @template T
 * @param {TopicIndex<T>} topicIndex
 * @param {string[][]} topics
 * @param {number} index
 */
function removeFromTopicIndex(topicIndex, topics, index) {
  if (topics.length !== 0) {
    const [first, ...rest] = topics;

    for (const topic of first) {
      if (topicIndex[topic]) {
        if (rest.length === 0) {
          const indices = topicIndex[topic].indices;
          const newIndices = indices.filter((i) => i !== index);
          topicIndex[topic].indices = newIndices;

          if (
            newIndices.length === 0 &&
            Object.keys(topicIndex[topic].next).length === 0
          ) {
            delete topicIndex[topic];
          }
        } else {
          removeFromTopicIndex(topicIndex[topic].next, rest, index);
        }
      }
    }
  }
}

/**
 * Finds callbacks whose topics intersect with the provided topics.
 * @template T
 * @param {TopicIndex<T>} topicIndex
 * @param {string[][]} topics
 * @param {TopicCallbacks<T>} callbacks
 * @returns {Callbacks<T>}
 */
function findMatchingCallbacks(topicIndex, topics, callbacks) {
  if (topics.length === 0) {
    return {};
  }

  /** @type {Callbacks<T>} */
  const matchingCallbacks = {};
  const [first, ...rest] = topics;

  for (const topic of first) {
    if (rest.length === 0) {
      if (topicIndex[topic]) {
        for (const index of topicIndex[topic].indices) {
          matchingCallbacks[index] = callbacks[index][1];
        }
      }
    } else {
      if (!topicIndex[topic]) {
        continue;
      }

      const next = findMatchingCallbacks(
        topicIndex[topic].next,
        rest,
        callbacks,
      );
      for (const [index, callback] of Object.entries(next)) {
        matchingCallbacks[index] = callback;
      }
    }
  }

  return matchingCallbacks;
}

// Stateless observer
// ==================

/**
 * Creates a new stateless observer.
 * @template T
 * @returns {Hub<T>}
 */
export function startStateless() {
  return {
    callbacks: {},
    index: 0,
  };
}

/**
 * Adds a callback to the stateless observer, returning the index.
 * @template T
 * @param {Hub<T>} hub
 * @param {Callback<T>} callback
 * @returns {number}
 */
export function addStateless(hub, callback) {
  hub.index++;
  hub.callbacks[hub.index] = callback;
  return hub.index;
}

/**
 * Invokes all callbacks synchronously with the given value.
 * @template T
 * @param {Hub<T>} hub
 * @param {T} value
 */
export function invokeStateless(hub, value) {
  invokeCallbacks(hub.callbacks, value);
}

/**
 * Removes a callback by its index.
 * @template T
 * @param {Hub<T>} hub
 * @param {number} index
 */
export function removeStateless(hub, index) {
  delete hub.callbacks[index];
}

/**
 * Clears all callbacks from the stateless observer.
 * @template T
 * @param {Hub<T>} hub
 */
export function stopStateless(hub) {
  hub.callbacks = {};
  hub.index = 0;
}

// Stateful observer
// =================

/**
 * Creates a new stateful observer with an initial state.
 * @template T
 * @param {T} value
 * @returns {StatefulHub<T>}
 */
export function startStateful(value) {
  return {
    callbacks: {},
    index: 0,
    value,
  };
}

/**
 * Adds a callback to the stateful observer, returning the current state and index.
 * @template T
 * @param {StatefulHub<T>} hub
 * @param {Callback<T>} callback
 * @returns {[T, number]}
 */
export function addStateful(hub, callback) {
  hub.index++;
  hub.callbacks[hub.index] = callback;
  return [hub.value, hub.index];
}

/**
 * Retrieves the current state.
 * @template T
 * @param {StatefulHub<T>} hub
 * @returns {T}
 */
export function currentState(hub) {
  return hub.value;
}

/**
 * Invokes all callbacks synchronously with a new state, updating the state.
 * @template T
 * @param {StatefulHub<T>} hub
 * @param {T} value
 */
export function invokeStateful(hub, value) {
  hub.value = value;
  invokeCallbacks(hub.callbacks, value);
}

/**
 * Removes a callback by its index.
 * @template T
 * @param {StatefulHub<T>} hub
 * @param {number} index
 */
export function removeStateful(hub, index) {
  delete hub.callbacks[index];
}

/**
 * Clears all callbacks and the value from the stateful observer.
 * @template T
 * @param {StatefulHub<T>} hub
 */
export function stopStateful(hub) {
  hub.callbacks = {};
  hub.index = 0;
  // @ts-ignore
  hub.value = undefined;
}

// Topic-based observer
// ====================

/**
 * Creates a new topic-based observer.
 * @template T
 * @returns {TopicBasedHub<T>}
 */
export function startTopicBased() {
  return {
    callbacks: {},
    index: 0,
    topicIndex: {},
  };
}

/**
 * Adds a callback with a list of topics to the topic-based observer, returning the index.
 * @template T
 * @param {TopicBasedHub<T>} hub
 * @param {string[][]} topics
 * @param {Callback<T>} callback
 * @returns {number}
 */
export function addTopicBased(hub, topics, callback) {
  topics = [...topics];

  hub.index++;
  hub.callbacks[hub.index] = [topics, callback];

  updateTopicIndex(hub.topicIndex, topics, hub.index);
  return hub.index;
}

/**
 * Invokes all matching callbacks synchronously with the given topics and value.
 * @template T
 * @param {TopicBasedHub<T>} hub
 * @param {string[][]} topics
 * @param {T} value
 */
export function invokeTopicBased(hub, topics, value) {
  topics = [...topics];

  const matchingCallbacks = findMatchingCallbacks(
    hub.topicIndex,
    topics,
    hub.callbacks,
  );
  invokeCallbacks(matchingCallbacks, value);
}

/**
 * Removes a callback by its index.
 * @template T
 * @param {TopicBasedHub<T>} hub
 * @param {number} index
 */
export function removeTopicBased(hub, index) {
  if (!hub.callbacks[index]) {
    return;
  }

  const topics = hub.callbacks[index][0];
  removeFromTopicIndex(hub.topicIndex, topics, index);
  delete hub.callbacks[index];
}

/**
 * Clears all callbacks and the topic index from the stateful observer.
 * @template T
 * @param {TopicBasedHub<T>} hub
 */
export function stopTopicBased(hub) {
  hub.callbacks = {};
  hub.index = 0;
  hub.topicIndex = {};
}

File: src/observer_ffi.erl
-module(observer_ffi).
-export([
    start_stateless/0,
    add_stateless/2,
    invoke_stateless/2,
    remove_stateless/2,
    stop_stateless/1,
    start_stateful/1,
    add_stateful/2,
    current_state/1,
    invoke_stateful/2,
    remove_stateful/2,
    stop_stateful/1,
    start_topic_based/0,
    add_topic_based/3,
    invoke_topic_based/3,
    remove_topic_based/2,
    stop_topic_based/1
]).

%% Helper functions
%% ================

%% Spawns processes to invoke each callback with the provided value,
%% and waits for all of them to complete.
spawn_invoke(Callbacks, Value) ->
    Monitors = lists:map(
        fun({_Index, Callback}) ->
            {_, Ref} = spawn_monitor(fun() -> Callback(Value) end),
            Ref
        end,
        maps:to_list(Callbacks)
    ),
    wait_for_monitors(Monitors).

%% Waits for all monitored processes to complete.
wait_for_monitors([]) ->
    ok;
wait_for_monitors([Ref | Rest]) ->
    receive
        {'DOWN', Ref, process, _, _} ->
            wait_for_monitors(Rest)
    end.

%% Updates the topic index with a new subscription.
update_topic_index(TopicIndex, Topics, Index) ->
    lists:foldl(
        fun(Topic, Acc) ->
            SubscriberIndices = maps:get(Topic, Acc, []),
            Acc#{Topic => [Index | SubscriberIndices]}
        end,
        TopicIndex,
        Topics
    ).

%% Removes a subscription from the topic index.
remove_from_topic_index(TopicIndex, Topics, Index) ->
    lists:foldl(
        fun(Topic, Acc) ->
            SubscriberIndices = maps:get(Topic, Acc, []),
            NewSubscriberIndices = lists:delete(Index, SubscriberIndices),
            if
                NewSubscriberIndices =:= [] ->
                    maps:remove(Topic, Acc);
                true ->
                    Acc#{Topic => NewSubscriberIndices}
            end
        end,
        TopicIndex,
        Topics
    ).

%% Finds callbacks whose topics intersect with the provided topics.
find_matching_callbacks(TopicIndex, Topics, Callbacks) ->
    SubscriberIndices = lists:flatmap(
        fun(Topic) ->
            maps:get(Topic, TopicIndex, [])
        end,
        Topics
    ),
    UniqueSubscriberIndices = lists:usort(SubscriberIndices),
    lists:foldl(
        fun(Index, Acc) ->
            {_, FoundCallback} = maps:get(Index, Callbacks),
            Acc#{Index => FoundCallback}
        end,
        #{},
        UniqueSubscriberIndices
    ).

%% Stateless observer
%% ==================

%% Starts the stateless observer process.
start_stateless() ->
    spawn(fun() -> stateless_loop(#{}, 0) end).

%% The main loop for the stateless observer.
stateless_loop(Callbacks, Index) ->
    receive
        {add, Callback, From} ->
            NewIndex = Index + 1,
            NewCallbacks = Callbacks#{NewIndex => Callback},
            From ! {ok, NewIndex},
            stateless_loop(NewCallbacks, NewIndex);
        {invoke, Value, From} ->
            spawn_invoke(Callbacks, Value),
            From ! {ok},
            stateless_loop(Callbacks, Index);
        {remove, Id} ->
            NewCallbacks = maps:remove(Id, Callbacks),
            stateless_loop(NewCallbacks, Index);
        stop ->
            ok
    end.

%% Adds a callback to the stateless observer, returning the index.
add_stateless(Process, Callback) ->
    Process ! {add, Callback, self()},
    receive
        {ok, Index} -> Index
    end.

%% Invokes all callbacks in parallel with the given value and waits for all of them to complete.
invoke_stateless(Process, Value) ->
    Process ! {invoke, Value, self()},
    receive
        {ok} -> ok
    end.

%% Removes a callback by its index.
remove_stateless(Process, Index) ->
    Process ! {remove, Index}.

%% Stops the stateless observer process.
stop_stateless(Process) ->
    Process ! stop.

%% Stateful observer
%% =================

%% Starts the stateful observer process with an initial state.
start_stateful(State) ->
    spawn(fun() -> stateful_loop(State, #{}, 0) end).

%% The main loop for the stateful observer.
stateful_loop(State, Callbacks, Index) ->
    receive
        {add, Callback, From} ->
            NewIndex = Index + 1,
            NewCallbacks = Callbacks#{NewIndex => Callback},
            From ! {ok, State, NewIndex},
            stateful_loop(State, NewCallbacks, NewIndex);
        {current, From} ->
            From ! {ok, State},
            stateful_loop(State, Callbacks, Index);
        {invoke, Value, From} ->
            spawn_invoke(Callbacks, Value),
            From ! {ok},
            stateful_loop(Value, Callbacks, Index);
        {remove, Id} ->
            NewCallbacks = maps:remove(Id, Callbacks),
            stateful_loop(State, NewCallbacks, Index);
        stop ->
            ok
    end.

%% Adds a callback to the stateful observer, returning the current state and index.
add_stateful(Process, Callback) ->
    Process ! {add, Callback, self()},
    receive
        {ok, State, Index} -> {State, Index}
    end.

%% Retrieves the current state.
current_state(Process) ->
    Process ! {current, self()},
    receive
        {ok, State} -> State
    end.

%% Invokes all callbacks in parallel with a new state, updating the state and waits for all callbacks to complete.
invoke_stateful(Process, Value) ->
    Process ! {invoke, Value, self()},
    receive
        {ok} -> ok
    end.

%% Removes a callback by its index.
remove_stateful(Process, Index) ->
    Process ! {remove, Index}.

%% Stops the stateful observer process.
stop_stateful(Process) ->
    Process ! stop.

%% Topic-based observer
%% ====================

%% Starts the topic-based observer process.
start_topic_based() ->
    spawn(fun() -> topic_based_loop(#{}, #{}, 0) end).

%% The main loop for the topic-based observer.
topic_based_loop(Callbacks, TopicIndex, Index) ->
    receive
        {add, Topics, Callback, From} ->
            NewIndex = Index + 1,
            NewCallbacks = Callbacks#{NewIndex => {Topics, Callback}},
            NewTopicIndex = update_topic_index(TopicIndex, Topics, NewIndex),
            From ! {ok, NewIndex},
            topic_based_loop(NewCallbacks, NewTopicIndex, NewIndex);
        {invoke, Topics, Value, From} ->
            MatchingCallbacks = find_matching_callbacks(TopicIndex, Topics, Callbacks),
            spawn_invoke(MatchingCallbacks, Value),
            From ! {ok},
            topic_based_loop(Callbacks, TopicIndex, Index);
        {remove, Id} ->
            case maps:get(Id, Callbacks, undefined) of
                {Topics, _} ->
                    NewTopicIndex = remove_from_topic_index(TopicIndex, Topics, Id),
                    NewCallbacks = maps:remove(Id, Callbacks),
                    topic_based_loop(NewCallbacks, NewTopicIndex, Index);
                undefined ->
                    topic_based_loop(Callbacks, TopicIndex, Index)
            end;
        stop ->
            ok
    end.

%% Adds a callback with topics to the topic-based observer, returning the index.
add_topic_based(Process, Topics, Callback) ->
    Process ! {add, Topics, Callback, self()},
    receive
        {ok, Index} -> Index
    end.

%% Invokes all matching callbacks in parallel with the given topics and value, and waits for all of them to complete.
invoke_topic_based(Process, Topics, Value) ->
    Process ! {invoke, Topics, Value, self()},
    receive
        {ok} -> ok
    end.

%% Removes a callback by its index.
remove_topic_based(Process, Index) ->
    Process ! {remove, Index}.

%% Stops the topic-based observer process.
stop_topic_based(Process) ->
    Process ! stop.
